### PR TITLE
fix(install/clean): infinite loops on circular dependencies

### PR DESCRIPTION
**Pull request by await-ovo**
add `resolved_packages`  memo cache to state to avoid infinite loops when installing circular dependencies

### CODE REVIEW

File: crates/package-manager/src/install_without_lockfile.rs [hunk 1]
```diff
@@ -9,6 +10,8 @@ use pacquet_registry::PackageVersion;
 use pacquet_tarball::MemCache;
 use pipe_trait::Pipe;
 
+pub type ResolvedPackages = MemoMap<String, bool>;
```
Comment (KSXGitHub) [inline crates/package-manager/src/install_without_lockfile.rs:13]: Please document what is the key and what is the value, what does `true` mean, and what does `false` mean.

File: crates/package-manager/src/install_without_lockfile.rs [hunk 2]
```diff
@@ -75,7 +81,18 @@ impl<'a> InstallWithoutLockfile<'a, ()> {
...
-        let InstallWithoutLockfile { tarball_mem_cache, http_client, config, .. } = self;
+        let InstallWithoutLockfile {
+            tarball_mem_cache,
+            http_client,
+            config,
+            resolved_packages,
+            ..
+        } = self;
+
+        if !resolved_packages.insert(package.to_virtual_store_name(), true) {
```
Comment (KSXGitHub) [inline crates/package-manager/src/install_without_lockfile.rs:92]: There's no `false`? Then I think `HashSet` might be better fit.

File: crates/cli/src/state.rs [hunk 3]
```diff
@@ -20,6 +21,8 @@ pub struct State {
     pub manifest: PackageManifest,
     /// Data from the `pnpm-lock.yaml` file.
     pub lockfile: Option<Lockfile>,
+    /// resolved packages
```
Comment (KSXGitHub) [inline crates/cli/src/state.rs:24]: This documentation is unhelpful. You may copy the (would be) documentation from the type alias to here.

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
Didn't expect you to fix this issue yourself. Very cool.

Anyway, I have some change requests.

File: crates/cli/tests/install.rs [hunk 4]
```diff
@@ -131,3 +131,28 @@ fn should_install_index_files() {
...
+        },
+    });
+    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");
+
+    eprintln!("Executing command...");
+    pacquet.with_arg("install").assert().success();
+
+    eprintln!("Snapshot");
+    let index_file_contents = index_file_contents(&store_dir);
+    insta::assert_yaml_snapshot!(index_file_contents);
```
Comment (zkochan) [inline crates/cli/tests/install.rs:155]: do really need this snapshot test? Maybe just check if the package is present in `node_modules`.

File: crates/package-manager/src/install_without_lockfile.rs (hunk 2 above)
Comment (await-ovo) [reply crates/package-manager/src/install_without_lockfile.rs]: Thanks for review, change it to dashset now.

File: crates/cli/tests/install.rs (hunk 4 above)
Comment (await-ovo) [reply crates/cli/tests/install.rs]: done ~

File: crates/package-manager/src/install_without_lockfile.rs [hunk 5]
```diff
@@ -9,6 +10,12 @@ use pacquet_registry::PackageVersion;
 use pacquet_tarball::MemCache;
 use pipe_trait::Pipe;
 
+/// In-memory cache for packages that have started resolving dependencies.
+///
+/// The contents of set is the package's virtual_store_name.
+/// e.g. @pnpm.e2e/dep-1@1.0.0 ->  @pnpm.e2e+dep-1@1.0.0
```
Comment (KSXGitHub) [inline crates/package-manager/src/install_without_lockfile.rs:16]:
```suggestion
/// e.g. `@pnpm.e2e/dep-1@1.0.0` →  `@pnpm.e2e+dep-1@1.0.0`
```

**Review by KSXGitHub — APPROVED** [review]

**Review by zkochan — APPROVED** [review]

**Pull request merged**